File: sdks/go/signing/manifestsignature.go
package signing

import (
	"bytes"
	"crypto/ed25519"
	"crypto/rand"
	"errors"
)

// SignatureEnvelope is §3's signature member: exactly the two string members protected
// and signature.
type SignatureEnvelope struct {
	Protected string
	Signature string
}

// canonicalizeOrWrap implements §10's wrapping rule. The underlying reason is one of
// canonical-json.md §9's closed set of five and travels ALONGSIDE the token rather than
// being reported as one of this document's ten, so a consumer switching on one never has
// to know about the other.
func canonicalizeOrWrap(manifest map[string]any) ([]byte, error) {
	canonical, err := CanonicalizeManifest(manifest)
	if err == nil {
		return canonical, nil
	}
	var reason *CanonicalizationError
	if errors.As(err, &reason) {
		return nil, &SignatureError{
			Reason:                 "canonicalization-failed",
			CanonicalizationReason: reason.Reason,
			Err:                    err,
		}
	}
	return nil, err
}

// decodeKeyOctets enforces §5's fixed length for a key member.
//
// A decode failure here is key-unsupported, never base64url-invalid: that token belongs
// to the ENVELOPE's two members (§8 step 2), and a key is not an envelope.
func decodeKeyOctets(value string) ([]byte, error) {
	octets, err := Base64URLDecode(value)
	if err != nil || len(octets) != ed25519.PublicKeySize {
		return nil, &SignatureError{Reason: "key-unsupported"}
	}
	return octets, nil
}

// GenerateSigningKey produces a fresh Ed25519 key pair as §5 JWKs.
//
// D is the 32-octet SEED, not Go's 64-octet ed25519.PrivateKey. RFC 8037 §2 defines d as
// the seed, and every other binding and every JOSE tool reads it that way; encoding the
// expanded private key here would produce a JWK nothing else can import. Measured:
// len(ed25519.NewKeyFromSeed(seed)) is 64 and len(priv.Seed()) is 32.
func GenerateSigningKey() (PrivateJWK, JWK, error) {
	public, private, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		return PrivateJWK{}, JWK{}, &SignatureError{Reason: "key-unsupported", Err: err}
	}
	encoded := Base64URLEncode(public)
	publicJWK := JWK{Kty: "OKP", Crv: "Ed25519", X: encoded}
	return PrivateJWK{JWK: publicJWK, D: Base64URLEncode(private.Seed())}, publicJWK, nil
}

// SignManifest implements §9. It does not mutate the manifest it was given.
func SignManifest(manifest map[string]any, key PrivateJWK) (SignatureEnvelope, error) {
	unsupported := &SignatureError{Reason: "key-unsupported"}

	// §9 step 1.
	if key.Kty != "OKP" || key.Crv != "Ed25519" {
		return SignatureEnvelope{}, unsupported
	}
	declaredPublic, err := decodeKeyOctets(key.X)
	if err != nil {
		return SignatureEnvelope{}, err
	}
	seed, err := decodeKeyOctets(key.D)
	if err != nil {
		return SignatureEnvelope{}, err
	}
	private := ed25519.NewKeyFromSeed(seed)

	// §9's correspondence rule, and it belongs HERE — at step 1, before step 4's
	// canonicalization. A key whose d does not correspond to its x produces an envelope
	// advertising a kid derived from x while carrying a signature made with d, so it can
	// never verify anywhere; left to the runtimes, one JavaScript engine accepts such a
	// pair, another rejects it at import, and Go ignores x entirely, since
	// NewKeyFromSeed derives the public key from the seed and never consults it.
	//
	// Checking late would make this binding answer canonicalization-failed where
	// TypeScript answers key-unsupported for one and the same (uncanonicalizable
	// manifest, non-corresponding key) input — §9's step list is not marked normative
	// the way §8's order is, so that divergence would have been invisible. Comparing the
	// derived public key to x directly is the direct route §9 permits; TypeScript signs
	// a probe and verifies it because Node cannot derive x from d.
	if !bytes.Equal(private.Public().(ed25519.PublicKey), declaredPublic) {
		return SignatureEnvelope{}, unsupported
	}

	// §9 step 2. §5's projection means a private key thumbprints as its own public half.
	kid, err := JWKThumbprint(key.JWK)
	if err != nil {
		return SignatureEnvelope{}, err
	}
	// §9 step 3.
	protected, err := EncodeProtectedHeader(ProtectedHeader{Alg: "EdDSA", Kid: kid})
	if err != nil {
		return SignatureEnvelope{}, err
	}
	// §9 step 4. CanonicalizeManifest clones, so the caller's map is untouched, and the
	// top-level signature member is what it strips — signing an already-signed manifest
	// therefore produces the same bytes as signing the unsigned one.
	canonical, err := canonicalizeOrWrap(manifest)
	if err != nil {
		return SignatureEnvelope{}, err
	}
	// §9 steps 5 and 6. PureEdDSA, RFC 8032 — no prehash of any kind.
	signature := ed25519.Sign(private, SigningInput(protected, canonical))
	return SignatureEnvelope{Protected: protected, Signature: Base64URLEncode(signature)}, nil
}

// VerifyManifestSignature implements §8's ten steps, in order, returning nil on success
// and a *SignatureError carrying exactly one §10 token otherwise.
//
// The order is normative, not advisory, and it is the most consequential thing in the
// contract. Every ordering verifies exactly the same set of VALID signatures, so no
// amount of round-trip testing distinguishes a conformant order from a non-conformant
// one; the orders differ only in which token an INVALID manifest reports. Do not reorder
// a check here even where a cheaper order gives the same answer on valid input.
//
// Synchronous, where TypeScript's is asynchronous, because crypto/ed25519 is.
func VerifyManifestSignature(manifest map[string]any, trusted []JWK) error {
	malformed := &SignatureError{Reason: "envelope-malformed"}

	// Step 1 — §3's envelope shape. publisher.id is required to be present and
	// well-formed so that a manifest which never named a publisher cannot present a
	// signature as though it had; its value is never compared against anything.
	publisher, isObject := manifest["publisher"].(map[string]any)
	if !isObject {
		return malformed
	}
	if id, isString := publisher["id"].(string); !isString || id == "" {
		return malformed
	}
	envelope, isObject := manifest["signature"].(map[string]any)
	if !isObject {
		return malformed
	}
	protectedMember, hasProtected := envelope["protected"].(string)
	signatureMember, hasSignature := envelope["signature"].(string)
	// Exactly means exactly: a third member — a header, a payload, a comment — is
	// malformed rather than merely unusual.
	if len(envelope) != 2 || !hasProtected || !hasSignature {
		return malformed
	}

	// Step 2 — BOTH members decode before either is parsed. A manifest whose protected
	// is valid base64url of malformed JSON and whose signature carries a '=' reports
	// base64url-invalid, not protected-malformed.
	protectedBytes, err := Base64URLDecode(protectedMember)
	if err != nil {
		return err
	}
	signatureBytes, err := Base64URLDecode(signatureMember)
	if err != nil {
		return err
	}

	// Steps 3 to 5.
	header, err := parseProtectedHeaderBytes(protectedBytes)
	if err != nil {
		return err
	}

	// Step 6 — thumbprintable keys only, and the rest are SKIPPED: a malformed entry in
	// a rotation set must not make every signature under that publisher unverifiable.
	// The skip is driven by JWKThumbprint's own verdict, which accepts any crv, and that
	// is what keeps step 7 reachable.
	var selected JWK
	found := false
	for _, candidate := range trusted {
		thumbprint, err := JWKThumbprint(candidate)
		if err != nil {
			continue
		}
		if thumbprint == header.Kid {
			selected, found = candidate, true
			break
		}
	}
	// An empty resolved key set lands here too.
	if !found {
		return &SignatureError{Reason: "kid-unknown"}
	}

	// Step 7 — reachable through exactly two routes: an OKP key on a curve other than
	// Ed25519, and an x that does not decode to 32 octets. X25519 is thumbprintable and
	// is a key-agreement curve rather than a signing one, which is why steps 6 and 7 are
	// two steps rather than one.
	// Kty is re-checked although step 6 only selects OKP keys: §8 states step 7 as
	// three conditions, and a verifier that leans on step 6's guarantee would silently
	// stop enforcing one of them if the selection rule ever moved.
	if selected.Kty != "OKP" || selected.Crv != "Ed25519" {
		return &SignatureError{Reason: "key-unsupported"}
	}
	publicKey, err := decodeKeyOctets(selected.X)
	if err != nil {
		return err
	}

	// Step 8 — the algorithm comes from the resolved key, never from the
	// attacker-supplied header, so alg is checked only now. The testable consequence is
	// that a manifest carrying both an unknown kid and a bogus alg reports kid-unknown.
	// An absent alg lands here too; §10 has no kid-missing counterpart for it.
	if header.Alg != "EdDSA" {
		return &SignatureError{Reason: "alg-unsupported"}
	}

	// Step 9. Last but one: every cheap structural check precedes both the expensive
	// serialization and the cryptographic operation, so a verifier does no
	// attacker-controlled work it can avoid.
	canonical, err := canonicalizeOrWrap(manifest)
	if err != nil {
		return err
	}

	// Step 10. The signing input incorporates the RECEIVED protected string verbatim, as
	// JWS specifies — re-encoding the header and comparing would reject a signature that
	// is cryptographically valid and make this contract unverifiable by any conformant
	// third-party JOSE signer.
	if len(signatureBytes) != ed25519.SignatureSize {
		return &SignatureError{Reason: "signature-invalid"}
	}
	if !ed25519.Verify(publicKey, SigningInput(protectedMember, canonical), signatureBytes) {
		return &SignatureError{Reason: "signature-invalid"}
	}
	return nil
}

File: sdks/typescript/src/signing/manifest-signature.ts
import { base64urlDecode, base64urlEncode } from "./base64url.js";
import { CanonicalizationError, canonicalizeManifest } from "./canonical-json.js";
import { SignatureError } from "./errors.js";
import { type Jwk, jwkThumbprint, type PrivateJwk } from "./jwk.js";
import { encodeProtectedHeader, parseProtectedHeaderBytes, signingInput } from "./jws.js";

/**
 * The detached JWS envelope, per `manifest-signature.md` §8 and §9.
 *
 * @moduleStability experimental
 *
 * Asynchronous because `crypto.subtle` is, which keeps this entry point runnable in a
 * browser, Deno or an edge worker. Go's binding is synchronous, so this is the minority
 * shape — the same two-against-one split `performHandshake` already carries.
 */
export interface ManifestSignatureEnvelope {
  readonly protected: string;
  readonly signature: string;
}

/**
 * §10. `canonicalization-failed` wraps `canonical-json.md` §9's closed set rather than
 * absorbing it, so the underlying reason travels alongside the token and neither set
 * grows by swallowing the other.
 */
function canonicalizeOrWrap(manifest: object): Uint8Array {
  try {
    return canonicalizeManifest(manifest);
  } catch (error) {
    if (error instanceof CanonicalizationError) {
      throw new SignatureError("canonicalization-failed", {
        canonicalizationReason: error.reason,
        cause: error,
      });
    }
    throw error;
  }
}

/**
 * `crypto.subtle` takes a `BufferSource`, which since TypeScript 5.7 means
 * `Uint8Array<ArrayBuffer>` specifically: a plain `Uint8Array` is
 * `Uint8Array<ArrayBufferLike>` and could in principle be backed by a `SharedArrayBuffer`,
 * which WebCrypto refuses. Copying is what proves it is not one, and at 32 or 64 octets —
 * or a manifest-sized signing input — the copy costs nothing next to the signature
 * operation it feeds. An assertion would be the alternative, and this module has none.
 */
function forCrypto(bytes: Uint8Array): Uint8Array<ArrayBuffer> {
  return new Uint8Array(bytes);
}

/**
 * §5 and §7/§9 step 1: a JWK's `x` — and a private JWK's `d` — must decode to exactly 32
 * octets. A decode failure here is `key-unsupported`, never `base64url-invalid`: that
 * token belongs to the *envelope*'s two members (§8 step 2), and a key is not an envelope.
 */
function decodeKeyOctets(value: string): Uint8Array<ArrayBuffer> {
  let bytes: Uint8Array;
  try {
    bytes = base64urlDecode(value);
  } catch {
    throw new SignatureError("key-unsupported");
  }
  if (bytes.length !== 32) throw new SignatureError("key-unsupported");
  return forCrypto(bytes);
}

export async function generateSigningKey(): Promise<{
  privateKey: PrivateJwk;
  publicKey: Jwk;
}> {
  // The algorithm is the bare string rather than `{ name: "Ed25519" }` so that the
  // `AlgorithmIdentifier` overload is the one selected: an object literal matches the
  // symmetric-key overload first and resolves to `CryptoKey`, which is not what Ed25519
  // returns. `in` then narrows the union without an assertion.
  const generated = await crypto.subtle.generateKey("Ed25519", true, ["sign", "verify"]);
  if (!("privateKey" in generated)) throw new SignatureError("key-unsupported");
  const exported = await crypto.subtle.exportKey("jwk", generated.privateKey);
  if (typeof exported.x !== "string" || typeof exported.d !== "string") {
    throw new SignatureError("key-unsupported");
  }
  return {
    privateKey: { kty: "OKP", crv: "Ed25519", x: exported.x, d: exported.d },
    publicKey: { kty: "OKP", crv: "Ed25519", x: exported.x },
  };
}

export async function signManifest(
  manifest: object,
  privateKey: PrivateJwk,
): Promise<ManifestSignatureEnvelope> {
  // §9 step 1.
  if (
    typeof privateKey !== "object" ||
    privateKey === null ||
    privateKey.kty !== "OKP" ||
    privateKey.crv !== "Ed25519" ||
    typeof privateKey.x !== "string" ||
    typeof privateKey.d !== "string"
  ) {
    throw new SignatureError("key-unsupported");
  }
  const publicKeyBytes = decodeKeyOctets(privateKey.x);
  decodeKeyOctets(privateKey.d);

  // §9 step 2. §5's projection means a private key thumbprints as its own public half.
  const kid = await jwkThumbprint(privateKey);
  // §9 step 3.
  const protectedB64 = encodeProtectedHeader({ alg: "EdDSA", kid });
  // §9 step 4.
  const canonical = canonicalizeOrWrap(manifest);
  // §9 step 5.
  const input = forCrypto(signingInput(protectedB64, canonical));

  try {
    const key = await crypto.subtle.importKey(
      "jwk",
      { kty: "OKP", crv: "Ed25519", x: privateKey.x, d: privateKey.d },
      { name: "Ed25519" },
      false,
      ["sign"],
    );
    const signature = new Uint8Array(await crypto.subtle.sign("Ed25519", key, input));

    // §9's correspondence rule. `kid` came from `x`, but the signature came from `d`; if
    // they disagree the envelope advertises a key that cannot verify it — anywhere, under
    // any implementation. bun accepts such a pair and signs with `d`, node rejects it at
    // importKey, so without this one binding has two answers depending on its runtime.
    // Deriving `x` from `d` is not portable (bun can, node cannot); signing a probe and
    // verifying it against the advertised `x` is.
    const verifier = await crypto.subtle.importKey(
      "raw",
      publicKeyBytes,
      { name: "Ed25519" },
      false,
      ["verify"],
    );
    if (!(await crypto.subtle.verify("Ed25519", verifier, signature, input))) {
      throw new SignatureError("key-unsupported");
    }
    // §9 step 6.
    return { protected: protectedB64, signature: base64urlEncode(signature) };
  } catch (error) {
    // §10's set is closed, and WebCrypto signals a rejected key by THROWING — measured, a
    // 31-octet raw key is a `DataError` in both bun and node. Every escape is normalized;
    // `cause` keeps the original for a debugger. A `SignatureError` is already in the set.
    if (error instanceof SignatureError) throw error;
    throw new SignatureError("key-unsupported", { cause: error });
  }
}

export async function verifyManifestSignature(
  manifest: object,
  trustedKeys: readonly Jwk[],
): Promise<void> {
  // Step 1 — the manifest itself, before any member is read. A corpus case can carry
  // `null` or a primitive, and reading `publisher` off `null` throws a raw TypeError that
  // escapes the closed token set entirely.
  if (typeof manifest !== "object" || manifest === null || Array.isArray(manifest)) {
    throw new SignatureError("envelope-malformed");
  }
  const document = manifest as Record<string, unknown>;

  // Step 1 — envelope shape.
  const publisher = document["publisher"];
  if (typeof publisher !== "object" || publisher === null || Array.isArray(publisher)) {
    throw new SignatureError("envelope-malformed");
  }
  const publisherId = (publisher as Record<string, unknown>)["id"];
  if (typeof publisherId !== "string" || publisherId === "") {
    throw new SignatureError("envelope-malformed");
  }
  const envelope = document["signature"];
  if (typeof envelope !== "object" || envelope === null || Array.isArray(envelope)) {
    throw new SignatureError("envelope-malformed");
  }
  const members = envelope as Record<string, unknown>;
  const protectedMember = members["protected"];
  const signatureMember = members["signature"];
  if (
    Object.keys(members).length !== 2 ||
    typeof protectedMember !== "string" ||
    typeof signatureMember !== "string"
  ) {
    throw new SignatureError("envelope-malformed");
  }

  // Step 2 — BOTH members decode before either is parsed. Decoding lazily is the natural
  // way to write this and reports `protected-malformed` where the contract says
  // `base64url-invalid`.
  const protectedBytes = base64urlDecode(protectedMember);
  const signatureBytes = base64urlDecode(signatureMember);

  // Steps 3-5.
  const header = parseProtectedHeaderBytes(protectedBytes);

  // Step 6 — thumbprintable keys only; a malformed entry in a rotation set must not make
  // every signature under that publisher unverifiable. `jwkThumbprint` rejects a non-OKP
  // key, so the skip has to be driven by its verdict rather than by a coarser type check
  // here — and it accepts any `crv`, which is what keeps step 7 reachable.
  let selected: Jwk | undefined;
  for (const candidate of trustedKeys) {
    let thumbprint: string;
    try {
      thumbprint = await jwkThumbprint(candidate);
    } catch (error) {
      // Only a rejection is a skip. A bug must still surface.
      if (error instanceof SignatureError) continue;
      throw error;
    }
    if (thumbprint === header.kid) {
      selected = candidate;
      break;
    }
  }
  if (selected === undefined) throw new SignatureError("kid-unknown");

  // Step 7 — X25519 is thumbprintable and is NOT a signing curve, which is why steps 6
  // and 7 are two steps rather than one.
  if (selected.kty !== "OKP" || selected.crv !== "Ed25519" || typeof selected.x !== "string") {
    throw new SignatureError("key-unsupported");
  }
  const publicKeyBytes = decodeKeyOctets(selected.x);

  // Step 8 — the algorithm comes from the resolved key, never from the attacker-supplied
  // header, so this is checked only now. An absent `alg` lands here too (§10 has no
  // `alg-missing`), which is why `ProtectedHeader.alg` is optional rather than literal.
  if (header.alg !== "EdDSA") throw new SignatureError("alg-unsupported");

  // Step 9.
  const canonical = canonicalizeOrWrap(document);

  // Step 10.
  if (signatureBytes.length !== 64) throw new SignatureError("signature-invalid");
  try {
    const key = await crypto.subtle.importKey("raw", publicKeyBytes, { name: "Ed25519" }, false, [
      "verify",
    ]);
    const ok = await crypto.subtle.verify(
      "Ed25519",
      key,
      forCrypto(signatureBytes),
      forCrypto(signingInput(protectedMember, canonical)),
    );
    if (!ok) throw new SignatureError("signature-invalid");
  } catch (error) {
    // Same closed-set normalization as `signManifest`, with the token this step owns.
    if (error instanceof SignatureError) throw error;
    throw new SignatureError("signature-invalid", { cause: error });
  }
}
